kvserver: allow logs from callbacks up to 15 replicas per updateReplicationGauges

wenyihu6 opened the pull request:
Part of: https://github.com/cockroachdb/cockroach/issues/151847
Epic: none

---

**kvserver: allow logs from callbacks up to 15 replicas per updateReplicationGauges**

Previously, logs from the decommission nudger were not gated by a vmodule and
could become spammy when many replicas were decommissioned at a low nudger
frequency. This commit introduces a per-store budget, allowing logs from
callbacks for up to 15 replicas per updateReplicationGauges call.

Drawbacks of this approach:
- Replicas are not visited in a sorted order, so we may be opening the floodgates
from 15 different replicas each iteration.
- Once a replica is permitted to log, its future logs from callbacks are not
  restricted.
- If EnqueueProblemRangeInReplicateQueueInterval is set too low, 1 and 2 may
  become worse.

For 1, we could consider visit the replica set with WithReplicasInOrder. I'm not
sure about the overhead here since updateReplicationGauges is called
periodically when collecting metrics.

Here are the reasons that I think this approach is acceptable for now:
- onEnqueueResult is unlikely to be reinvoked for replicas already in the queue
unless they are processing or in purgatory (both are short-lived states we want
visibility into). Once processed, replicas are removed from the set.
onProcessResult should be called at most twice. For replicas merely waiting in
the queue, the callback is not invoked, since their priority should not be
actively updated.
- We could cap logging per maybeEnqueueProblemRange, but granting full logging
permission for each replica simplifies reasoning and gives complete visibility
for specific replias.
- In practice, escalations show that slow decommissioning usually involves <15
  ranges, and EnqueueProblemRangeInReplicateQueueInterval is typically large
  (~15 minutes).

---

**kvserver: rename shouldLog to maybeLog and change vlevel to a var**

This commit renames shouldLog to maybeLog in maybeEnqueueProblemRange and
refactors vmodule logging level to use a dynamic variable instead of an if
statement to choose between Info and VInfo.

wenyihu6 committed:
kvserver: allow logs from callbacks up to 15 replicas per updateReplicationGauges

Previously, logs from the decommission nudger were not gated by a vmodule and
could become spammy when many replicas were decommissioned at a low nudger
frequency. This commit introduces a per-store budget, allowing logs from
callbacks for up to 15 replicas per updateReplicationGauges call.

Drawbacks of this approach:
- Replicas are not visited in a sorted order, so we may be opening the floodgates
from 15 different replicas each iteration.
- Once a replica is permitted to log, its future logs from callbacks are not
  restricted.
- If EnqueueProblemRangeInReplicateQueueInterval is set too low, 1 and 2 may
  become worse.

For 1, we could consider visit the replica set with WithReplicasInOrder. I'm not
sure about the overhead here since updateReplicationGauges is called
periodically when collecting metrics.

Here are the reasons that I think this approach is acceptable for now:
- onEnqueueResult is unlikely to be reinvoked for replicas already in the queue
unless they are processing or in purgatory (both are short-lived states we want
visibility into). Once processed, replicas are removed from the set.
onProcessResult should be called at most twice. For replicas merely waiting in
the queue, the callback is not invoked, since their priority should not be
actively updated.
- We could cap logging per maybeEnqueueProblemRange, but granting full logging
permission for each replica simplifies reasoning and gives complete visibility
for specific replias.
- In practice, escalations show that slow decommissioning usually involves <15
  ranges, and EnqueueProblemRangeInReplicateQueueInterval is typically large
  (~15 minutes).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/152885)
<!-- Reviewable:end -->

tbg reviewed (approved):
> @tbg reviewed 1 of 3 files at r2, 2 of 5 files at r3, 1 of 2 files at r4, 1 of 2 files at r5, 1 of 2 files at r7, 1 of 1 files at r9, 2 of 2 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152885)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/store.go` line 3497 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152885#-OZDobOU8JGRB30i3kO1:-OZDobOV2hs8eQPMuN2_:bacozmj) ([raw file](https://github.com/cockroachdb/cockroach/blob/08e54db551ccb3dc948593ecc1b6e05f9bfedbeb/pkg/kv/kvserver/store.go#L3497)):*
> ```Go
> 				// NB: Enqueue is disabled by default from here and throttled async if
> 				// enabled.
> 				shouldLog := logBudgetOnDecommissioningNudger > 0
> ```

nit: shouldLog -> mayLog since mostly `maybeEnqueueProblemRange` mostly doesn't log even if it's allowed to (due to the cluster setting check).
___
*[`pkg/kv/kvserver/replica.go` line 2965 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152885#-OZDoseV0ctIXJQ8LRs6:-OZDoseV0ctIXJQ8LRs7:b-o2b0e0) ([raw file](https://github.com/cockroachdb/cockroach/blob/08e54db551ccb3dc948593ecc1b6e05f9bfedbeb/pkg/kv/kvserver/replica.go#L2965)):*
> ```Go
> 		allocatorimpl.AllocatorReplaceDecommissioningVoter.Priority(), processCallback{
> 			onEnqueueResult: func(indexOnHeap int, err error) {
> 				if shouldLog {
> ```

Couldn't this be made much denser like so:

```
level := 2
if shouldLog { level = 0 }

log.VInfof(ctx, level, ...)
```


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
kvserver: rename shouldLog to maybeLog and change vlevel to a var

This commit renames shouldLog to maybeLog in maybeEnqueueProblemRange and
refactors vmodule logging level to use a dynamic variable instead of an if
statement to choose between Info and VInfo.

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152885)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/kv/kvserver/replica.go` line 2965 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152885#-OZDoseV0ctIXJQ8LRs6:-OZFWN200WyO0ked3XHc:b1hzddp) ([raw file](https://github.com/cockroachdb/cockroach/blob/08e54db551ccb3dc948593ecc1b6e05f9bfedbeb/pkg/kv/kvserver/replica.go#L2965)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Couldn't this be made much denser like so:

```
level := 2
if shouldLog { level = 0 }

log.VInfof(ctx, level, ...)
```
</blockquote></details>

Done, going to rebase on master. 
___
*[`pkg/kv/kvserver/store.go` line 3497 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152885#-OZDobOU8JGRB30i3kO1:-OZFWOwwBhMbbUCzERIr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/08e54db551ccb3dc948593ecc1b6e05f9bfedbeb/pkg/kv/kvserver/store.go#L3497)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

nit: shouldLog -> mayLog since mostly `maybeEnqueueProblemRange` mostly doesn't log even if it's allowed to (due to the cluster setting check).
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
